Documentation/design-docs/native-hosting-initialize.md

```diff
@@ -0,0 +1,182 @@
+    int argc,
+    const char_t * argv[],
+    const char_t * app_path  
+    const hostfxr_initialize_parameters * parameters,
+    hostfxr_handle * handle
+);
+```
+
+Initializes the host for running a managed application. Also used for running SDK commands.
+When used to execute an app, the `app_path` (or CLI equivalent) will be used to locate the `.runtimeconfig.json` and the `.deps.json` which will be used to load the application and its dependent frameworks.
+* `argc` and `argv` - the command line - entirely optional
+* `app_path` - path to the application (the managed `.dll`) to run. This can be `nullptr` if the app is specified in the command line arguments or if it's an SDK command.
+* `parameters` - additional parameters - see `hostfxr_initialize_parameters` for details. (Could be made optional potentially)
+* `handle` - output parameter. On success receives an opaque value which identifies the initialized host. The handle should be closed by calling `hostfxr_shutdown`.
+
+This function can only be called once per-process. It's not supported to run multiple apps in one process (even sequentially).
+
+*Note: This is effectively a replacement for the `hostfxr_main_startupinfo` and `hostfxr_main` - both of those APIs should be possible to implement using the `hostfxr_initialize_for_app` and the `hostfxr_run_app`.*
+
+
+``` C++
+int hostfxr_initialize_for_runtime_config(
+    const char_t * runtime_config_path
+    const hostfxr_initialize_parameters * parameters,
+    hostfxr_handle * handle
+);
+```
+
+This function would load the specified `.runtimeconfig.json`, resolve all frameworks and resolve all the assets from those frameworks. Then prepare runtime initialization where the TPA contains only frameworks. Note that this case does NOT consume any `.deps.json` from the app/component (only processes the framework's `.deps.json`). This entry point is intended for `comhost`/`ijwhost`/`nethost` and similar scenarios.
+* `runtime_config_path` - path to the `.runtimeconfig.json` file to process. Unlike with the `hostfxr_initialize_for_app`, there is no `.deps.json` from the app/component which will be processed.
+* `parameters` - additional parameters - see `hostfxr_initialize_parameters` for details. (Could be made optional potentially)
+* `handle` - output parameter. On success receives an opaque value which identifies the initialized host. The handle should be closed by calling `hostfxr_shutdown`.
+
+This function can be called multiple times in a process.
+* If it's called when no runtime is present, it will run through the steps to "initialize" the runtime (so resolving frameworks and so on).
+* If it's called when there already is CoreCLR in the process (loaded through the `hostfxr`, direct usage of `coreclr` is not supported), then the function determines if the specified runtime configuration is compatible with the existing runtime and frameworks. If it is, it returns a valid handle, otherwise it fails.
```

Comment:
That's an interesting point - I think the answer is no - one of the initialize calls would fail. Due to some limitations we can't jsut look for CoreCLR in the process, instead we rely on global variable in the hosting components. This is something we may need to implement - I know that `comhost` introduces some parts of this, but we may need to extend it even more broadly.

But the general point is, `hostfxr` will have a way to track the "loaded runtime". We could make it so that initializing the host == marking the runtime as loaded (even though we haven't really loaded CoreCLR yet, but we did load `hostpolicy` which is effectively part of the runtime).

Reply:
I added a detailed description - we will have to look into this some more during the implementation. COM activation effectively needs to be able to initialize/run from multiple threads at the same time - so we need to have a solid mechanism in place.